### PR TITLE
doc:  add a generic section on exploring package sets

### PR DESCRIPTION
**Pull request by fricklerhandwerk**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: doc/languages-frameworks/haskell.section.md [hunk 1]
```diff
@@ -25,21 +25,12 @@ All of these packages are originally defined in the `haskellPackages` package
...
+The `haskellPackages` set includes at least one version of every package from [Hackage](https://hackage.haskell.org/) as well as some manually injected packages.
+Haskell packages can be searched on [search.nixos.org](https://search.nixos.org/packages?query=haskellPackages) or listed on the command line with [`nix-instantiate`](https://nixos.org/manual/nix/stable/command-ref/nix-instantiate):
 
-```console
-$ nix-env -f '<nixpkgs>' -qaP -A haskellPackages
-haskellPackages.a50                                                         a50-0.5
-haskellPackages.AAI                                                         AAI-0.2.0.1
-haskellPackages.aasam                                                       aasam-0.2.0.0
-haskellPackages.abacate                                                     abacate-0.0.0.0
-haskellPackages.abc-puzzle                                                  abc-puzzle-0.2.1
```
Comment (Atemu) [inline doc/languages-frameworks/haskell.section.md:39]: The new method does not show version information which might be critical.

It'd be great to retain the example output in any case though.

**Review by Atemu — APPROVED** [review]

**Review by sternenseemann — DISMISSED** [review]
This just seems silly to me. There's nothing wrong with using `nix-env` in this way and the presented alternative is strictly worse?

**Comment by Atemu** [comment]
Listing packages given a (possibly nested) attrset while ignoring any eval errors is one of the only sensible uses of nix-env. What we want to get away from is using nix-env for imperative package management because it really sucks quite a lot at that.

**Comment by fricklerhandwerk** [comment]
@Atemu would I be missing something crucial if I claim this can be done in the 4-liner I added as an example?

**Comment by Atemu** [comment]
Yes, it'd fall on its nose on the first attribute that doesn't eval or doesn't have a version. haskellPackages is quite well behaved in this regard but go I dare you to this on `pkgs`... The only way to get this sort of information reliably on `pkgs` is `nix-env`; I've been there.

**Comment by fricklerhandwerk** [comment]
Incredible...

https://github.com/NixOS/nix/blob/master/src/nix-env/nix-env.cc#L195
https://github.com/NixOS/nix/blob/master/src/libexpr/get-drvs.cc#L311-L330

How about we all forget that code ever existed, merge this PR, and focus on something else?

**Comment by Atemu** [comment]
I don't have a strong opinion on this and I don't maintain this part of Nixpkgs. I also don't have anything more to contribute to this discussion, so I'm going to unsubscribe here.

**Comment by fricklerhandwerk** [comment]
@maralorn @ncfavier opinions?

**Comment by maralorn** [comment]
I have no strong opinion on this and I frankly don’t feel like arbitrating this on my free time. So I will just leave some thoughts:

I empathize with the desire to get rid of references to nix-env. I mean in general even mentioning that packages won’t get listed by `nix-env -qa` seems quite irrelevant because that workflow seems rather alien to me. I wonder if anyone ever actually needs this on the command line.

On the other hand the fact that this command becomes so much more complicated without it is annoying, eval errors will slip in from time to time and this usage of nix-env seems harmless in contrast to the usual critique of that command.

 I think the biggest advantage of having the listing in the manual is that it makes the package set a bit more graspable.

My usual workflow is search.nixos.org or nix repl + tabcompletion, I don‘t know if that’s representative but if it is it might be the way to go also for the manual …

**Comment by sternenseemann** [comment]
> Incredible... just ignore assertions?

You fundamentally have to do this when dealing with how nixpkgs works. `haskellPackages` is a *very* tame corner of nixpkgs, but even here you'll need to deal with assertions if you want to know `outPath`s.

> the only stable, reliable, and (mostly) sane interface to Nix is the Nix language.

It's frankly bizarre to state that to list packages you should

1. Install `jq`,
2. Copy and paste some code snippet that is specific to `haskellPackages`

instead of using a subcommand of `nix-env(1)` that is not problematic.

> I thought it was widely agreed-upon that using `nix-env` should not be encouraged

Yes, `nix-env` for package management is to be discouraged, but that concerns `nix-env -u` and `nix-env -i`. I just don't see any alternative to `nix-env -qP` in this case.

**Comment by maralorn** [comment]
I like this in general.

I still think that having a few example packages listed in the Haskell section makes it easier to quickly grasp the structure of the package sets in question. But overall this change is already an improvement.

**Pull request merged**